File: BitTorrent/UDP Networking/InternetProtocol.m
//
//  InternetProtocol.m
//  BitTorrent
//
//  Created by Ben Davis on 16/11/2014.
//  Copyright (c) 2014 Ben Davis. All rights reserved.
//

#import "InternetProtocol.h"
#include <ifaddrs.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <netdb.h>

#define IOS_CELLULAR    @"pdp_ip0"
#define IOS_WIFI        @"en0"
#define IP_ADDR_IPv4    @"ipv4"
#define IP_ADDR_IPv6    @"ipv6"

@implementation InternetProtocol

+ (NSString *)getIPAddress:(BOOL)preferIPv4 {
    NSArray *searchArray = preferIPv4 ?
    @[ IOS_WIFI @"/" IP_ADDR_IPv4, IOS_WIFI @"/" IP_ADDR_IPv6, IOS_CELLULAR @"/" IP_ADDR_IPv4, IOS_CELLULAR @"/" IP_ADDR_IPv6 ] :
    @[ IOS_WIFI @"/" IP_ADDR_IPv6, IOS_WIFI @"/" IP_ADDR_IPv4, IOS_CELLULAR @"/" IP_ADDR_IPv6, IOS_CELLULAR @"/" IP_ADDR_IPv4 ] ;
    
    NSDictionary *addresses = [self getIPAddresses];
    NSLog(@"addresses: %@", addresses);
    
    __block NSString *address;
    [searchArray enumerateObjectsUsingBlock:^(NSString *key, NSUInteger idx, BOOL *stop)
     {
         address = addresses[key];
         if(address) *stop = YES;
     } ];
    return address ? address : @"0.0.0.0";
}

+ (NSDictionary *)getIPAddresses {
    NSMutableDictionary *addresses = [NSMutableDictionary dictionaryWithCapacity:8];
    
    // retrieve the current interfaces - returns 0 on success
    struct ifaddrs *interfaces;
    if(!getifaddrs(&interfaces)) {
        // Loop through linked list of interfaces
        struct ifaddrs *interface;
        for(interface=interfaces; interface; interface=interface->ifa_next) {
            if(!(interface->ifa_flags & IFF_UP) /* || (interface->ifa_flags & IFF_LOOPBACK) */ ) {
                continue; // deeply nested code harder to read
            }
            const struct sockaddr_in *addr = (const struct sockaddr_in*)interface->ifa_addr;
            char addrBuf[ MAX(INET_ADDRSTRLEN, INET6_ADDRSTRLEN) ];
            if(addr && (addr->sin_family==AF_INET || addr->sin_family==AF_INET6)) {
                NSString *name = [NSString stringWithUTF8String:interface->ifa_name];
                NSString *type;
                if(addr->sin_family == AF_INET) {
                    if(inet_ntop(AF_INET, &addr->sin_addr, addrBuf, INET_ADDRSTRLEN)) {
                        type = IP_ADDR_IPv4;
                    }
                } else {
                    const struct sockaddr_in6 *addr6 = (const struct sockaddr_in6*)interface->ifa_addr;
                    if(inet_ntop(AF_INET6, &addr6->sin6_addr, addrBuf, INET6_ADDRSTRLEN)) {
                        type = IP_ADDR_IPv6;
                    }
                }
                if(type) {
                    NSString *key = [NSString stringWithFormat:@"%@/%@", name, type];
                    addresses[key] = [NSString stringWithUTF8String:addrBuf];
                }
            }
        }
        // Free memory
        freeifaddrs(interfaces);
    }
    return [addresses count] ? addresses : nil;
}

+(NSString *)ipAddressOfHostname:(NSString *)hostName {
    struct hostent *host_entry = gethostbyname([hostName cStringUsingEncoding:NSASCIIStringEncoding]);
    char *buff;
    if (host_entry != NULL) {
        buff = inet_ntoa(*((struct in_addr *)host_entry->h_addr_list[0]));
        NSString *ip = [NSString stringWithUTF8String:buff];
        return ip;
    } else {
        return nil;
    }
}

+ (NSString*)ipAddressFromSockAddrData:(NSData*)data {
    struct sockaddr_in * socketAddress = (struct sockaddr_in *)data.bytes;
    struct	in_addr ipAsStruct = ((struct sockaddr_in)*(socketAddress)).sin_addr;
    char *buff;
    buff = inet_ntoa(ipAsStruct);
    NSString *ip = [NSString stringWithUTF8String:buff];
    return ip;
}

+ (uint16_t)portFromSockAddrData:(NSData*)data {
    struct sockaddr_in * socketAddressPtr = (struct sockaddr_in *)data.bytes;
    struct sockaddr_in socketAddress = ((struct sockaddr_in)*(socketAddressPtr));
    return socketAddress.sin_port;
}

@end

File: BitTorrent/UDP Networking/InternetProtocol.swift
//
//  InternetProtocol.swift
//  BitTorrent
//
//  Created by Ben Davis on 03/07/2017.
//  Copyright © 2017 Ben Davis. All rights reserved.
//

import Foundation

struct InternetProtocol {
    
    static let IOS_CELLULAR_INTERFACE_NAME = "pdp_ip0"
    static let IOS_WIFI_INTERFACE_NAME = "en0"
    static let IP_ADDR_IPv4_INTERFACE_NAME = "ipv4"
    static let IP_ADDR_IPv6_INTERFACE_NAME = "ipv6"
    
    static func ipAddress(fromSockAddrData data: Data) -> String? {
        let socketAddress = data.withUnsafeBytes() { (pointer: UnsafePointer<sockaddr_in>) in
            return pointer.pointee
        }
        guard let resultCString = inet_ntoa(socketAddress.sin_addr) else {
            return nil
        }
        return String(cString: resultCString)
    }
    
    static func port(fromSockAddrData data: Data) -> UInt16 {
        let socketAddress = data.withUnsafeBytes() { (pointer: UnsafePointer<sockaddr_in>) in
            return pointer.pointee
        }
        return socketAddress.sin_port
    }
    
    static func getIPAddress(of hostname: String) -> String? {
        
        guard let hostnameCString = hostname.cString(using: .ascii),
            let hostEntry = gethostbyname(hostnameCString)?.pointee,
            let hostAddressList = hostEntry.h_addr_list?.pointee else {
                return nil
        }
        
        let firstHostAddress = hostAddressList.withMemoryRebound(to: in_addr.self, capacity: 1) { $0.pointee }
        let firstHostAddressCString = inet_ntoa(firstHostAddress)!
        return String(cString: firstHostAddressCString)
    }
    
    static func getLocalIPAddress(preferIPv4: Bool = true) -> String? {
        
        // Prefer wifi over cellular
        let searchArray = [
            ifaddrs.nameAndTypeString(from: IOS_WIFI_INTERFACE_NAME, isIpv4: preferIPv4),
            ifaddrs.nameAndTypeString(from: IOS_WIFI_INTERFACE_NAME, isIpv4: !preferIPv4),
            ifaddrs.nameAndTypeString(from: IOS_CELLULAR_INTERFACE_NAME, isIpv4: preferIPv4),
            ifaddrs.nameAndTypeString(from: IOS_CELLULAR_INTERFACE_NAME, isIpv4: !preferIPv4),
            ]
        
        let addresses = getLocalIPAddresses()
        
        for searchItem in searchArray {
            if let result = addresses[searchItem] {
                return result
            }
        }
        
        return nil
    }
    
    static func getLocalIPAddresses() -> [String: String] {
        
        var addresses: [String: String] = [:]
        
        // Get list of all network interfaces on the local machine:
        var ifaddrsPointer : UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddrsPointer) == 0, let ifaddrsList = ifaddrsPointer?.pointee else {
            return [:]
        }
        
        // For each network interface ...
        for ifaddrs in ifaddrsList {
            if !ifaddrs.isUpAndRunning || ifaddrs.isLoopbackNet {
                continue
            }
            
            if ifaddrs.isIpv4 || ifaddrs.isIpv6 {
                if let addressString = ifaddrs.convertToIPString(), let name = ifaddrs.nameAndTypeString() {
                    addresses[name] = addressString
                }
            }
        }
        
        freeifaddrs(ifaddrsPointer)
        
        return addresses
    }
}


public class ifaddrsIterator: IteratorProtocol {
    public typealias Element = ifaddrs
    
    var nextElement: ifaddrs?
    
    init(first: ifaddrs) {
        nextElement = first
    }
    
    public func next() -> ifaddrs? {
        let result = nextElement
        nextElement = result?.ifa_next?.pointee
        return result
    }
}

extension ifaddrs: Sequence {
    public func makeIterator() -> ifaddrsIterator {
        return ifaddrsIterator(first: self)
    }
}

extension ifaddrs {
    
    var name: String {
        return String(cString: ifa_name)
    }
    
    var isIpv4: Bool {
        let addr = ifa_addr.pointee
        return addr.sa_family == UInt8(AF_INET)
    }
    
    var isIpv6: Bool {
        let addr = ifa_addr.pointee
        return addr.sa_family == UInt8(AF_INET6)
    }
    
    var isUpAndRunning: Bool {
        let flags = Int32(ifa_flags)
        let upAndRunningFlags = (IFF_UP|IFF_RUNNING)
        return (flags & upAndRunningFlags) == upAndRunningFlags
    }
    
    var isLoopbackNet: Bool {
        let flags = Int32(ifa_flags)
        return (flags & IFF_LOOPBACK) == IFF_LOOPBACK
    }
    
    func convertToIPString() -> String? {
        return ifa_addr.pointee.toString()
    }
    
    func nameAndTypeString() -> String? {
        
        guard isIpv4 || isIpv6 else {
            return nil
        }
        
        return ifaddrs.nameAndTypeString(from: name, isIpv4: isIpv4)
    }
    
    static func nameAndTypeString(from name: String, isIpv4: Bool) -> String {
        return name + "/" + (isIpv4 ?
            InternetProtocol.IP_ADDR_IPv4_INTERFACE_NAME :
            InternetProtocol.IP_ADDR_IPv6_INTERFACE_NAME)
    }
}

extension sockaddr {
    
    func toString() -> String? {
        
        var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
        var addr = self
        
        guard getnameinfo(&addr,
                          socklen_t(addr.sa_len),
                          &hostname,
                          socklen_t(hostname.count),
                          nil,
                          socklen_t(0),
                          NI_NUMERICHOST) == 0 else { return nil }
        
        return String(validatingUTF8: hostname)
    }
}
